pkg/terraform/errors/errors.go

```diff
@@ -16,40 +16,169 @@ limitations under the License.
 
 package errors
 
+import (
+	"fmt"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
+)
+
+const (
+	levelError = "error"
+)
+
+type tfError struct {
+	message string
+}
+
 type applyFailed struct {
-	log string
+	*tfError
+}
+
+// TerraformLog represents relevant fields of a Terraform CLI JSON-formatted log line
+type TerraformLog struct {
+	Level      string        `json:"@level"`
+	Message    string        `json:"@message"`
+	Diagnostic LogDiagnostic `json:"diagnostic"`
+}
+
+// LogDiagnostic represents relevant fields of a Terraform CLI JSON-formatted
+// log line diagnostic info
+type LogDiagnostic struct {
+	Severity string `json:"severity"`
+	Summary  string `json:"summary"`
+	Detail   string `json:"detail"`
+	Range    Range  `json:"range"`
+}
+
+// Range represents a line range in a Terraform workspace file
+type Range struct {
+	FileName string `json:"filename"`
+}
+
+func (t *tfError) Error() string {
+	return t.message
 }
 
-func (a *applyFailed) Error() string {
-	return a.log
+func newTFError(message string, logs []byte) (string, *tfError) {
```

Comment:
nitpitck
```suggestion
func newTFError(message string, logs []byte) (*tfError, error) {
```

Reply:
Because `tfError` is shared (embedded) inside actual error types and because we need to consume the returned error instance as a `tfError`, doing this will necessitate type cases at those sites. I'd prefer to keep it as it's as it's not part of the public interface.